### PR TITLE
Refactor codebase to apply selected ruff linting and formatting rules

### PR DESCRIPTION
**Pull request by NKeleher**
I've gone through the full codebase to apply the following ruff linting rules:

```
select = [
  "F", # Pyflakes
  "E", # pycodestyle errors
  "W", # pycodestyle warnings
  "I", # isort
  "UP", # pyupgrade
  "SIM", # flake8-simplify
  "TRY", # tryceratops
]
```

I disabled the [SIM110](https://docs.astral.sh/ruff/rules/reimplemented-builtin/
) rule because that conflicts with numba.

I disabled some rules inline, most specifically [F841](https://docs.astral.sh/ruff/rules/unused-variable/) because I wasn't sure if future modifications would need these objects. Look for  `# noqa: F841` in the code.

I have added ruff to the github workflow to lint code. I'm leaving the black linting as it follows _after_ruff to stay consistent with the previous linting. However, black _could_ be removed if there is a preference to do all linting and formatting with black.

Still to do is apply the ruff "D" linting rules for docstyle.

### CODE REVIEW

**Comment by NKeleher** [comment]
Still a WIP, need to fix the following failing tests:

```
================================================================================ short test summary info =================================================================================
FAILED tests/test_did.py::test_event_study - KeyError: False
FAILED tests/test_did.py::test_did2s - KeyError: False
FAILED tests/test_did.py::test_errors - KeyError: False
FAILED tests/test_vs_fixest.py::test_singleton_dropping - AssertionError: 
========================================================== 4 failed, 1308 passed, 93 skipped, 144 warnings in 737.70s (0:12:17) ==========================================================
```

**Comment by s3alfisc** [comment]
> Still a WIP, need to fix the following failing tests:
Wow, that was really fast! Should I take a look at the failing tests tonight?

File: pyfixest/did/did2s.py [hunk 1]
```diff
@@ -126,37 +128,34 @@ def _did2s_estimate(
...
+            if data[treatment].nunique() != 2:
+                raise ValueError(f"The treatment variable {treatment} must be boolean.")
+            data[treatment] = data[treatment].astype(bool)
+            warnings.warn(
+                f"The treatment variable {treatment} was converted to boolean."
+            )
+        _not_yet_treated_data = data[data[treatment] is False]
     else:
-        _not_yet_treated_data = data[data["ATT"] == False]
+        _not_yet_treated_data = data[data["ATT"] is False]
```
Comment (s3alfisc) [inline pyfixest/did/did2s.py:158]: @NKeleher this is where the KeyError for the `did2s` test stem from.

Comment (s3alfisc) [reply pyfixest/did/did2s.py:158]: In fact I believe it's the same error in all of these: 

FAILED tests/test_did.py::test_event_study - KeyError: False -> also tests the did2s implementation
FAILED tests/test_did.py::test_did2s - KeyError: False -> error in did2s
FAILED tests/test_did.py::test_errors - KeyError: False -> error in error checking for did2s

**Comment by s3alfisc** [comment]
> FAILED tests/test_vs_fixest.py::test_singleton_dropping - AssertionError: 

This one stems from [this line](https://github.com/s3alfisc/pyfixest/blob/4fd321aaf62d9ede1f04738a5d03c83585dbdeed/tests/test_vs_fixest.py#L824): `fixest` drops one singleton fixed effects and three NaNs, `pyfixest` now somehow does not drop the singleton fixed effect.

**Comment by NKeleher** [comment]
Thanks for the pointers on those test failures. I can take a closer look at resolving last this evening.

**Comment by NKeleher** [comment]
Thanks again for pointing me to the potential source of the test errors @s3alfisc. I've resolved the errors and all tests are now passing: 
image

Let me know if you'd like any further changes. I'm going to move to work on the linting of the docstrings for the docs by adding in the "D" linting rules for ruff. I can continue to work on this branch or, if you merge, this branch to master, I can start on a new branch. Just let me know which you prefer.

**Comment by NKeleher** [comment]
> > FAILED tests/test_vs_fixest.py::test_singleton_dropping - AssertionError:
> 
> This one stems from [this line](https://github.com/s3alfisc/pyfixest/blob/4fd321aaf62d9ede1f04738a5d03c83585dbdeed/tests/test_vs_fixest.py#L824): `fixest` drops one singleton fixed effects and three NaNs, `pyfixest` now somehow does not drop the singleton fixed effect.

This ended up being the fix for the singletons error:
https://github.com/s3alfisc/pyfixest/pull/311/files#diff-bebbce8c57299b3c2250d91104c354211f0128803da499e317a8e093bf4fc16fR291

**Comment by s3alfisc** [comment]
>  I've resolved the errors and all tests are now passing:

Very cool, I'll start my review then! 

> I can continue to work on this branch or, if you merge, this branch to master, I can start on a new branch. Just let me know which you prefer.

I think it would be best if I were to merge this master already, don't you think? Else the next PR would get very big & additionally I will already get to apply the new linting rules for all new code that I write.

**Comment by NKeleher** [comment]
I agree that merging to the master branch, after you've reviewed will make further revisions easier. Sounds great!

File: pyfixest/utils.py [hunk 2]
```diff
@@ -35,80 +36,101 @@ def ssc(adj=True, fixef_k="none", cluster_adj=True, cluster_df="min"):
...
+
+    Raises
+    ------
+    ValueError
+        If vcov_type is not "iid", "hetero", or "CRV", or if cluster_df is neither "conventional" nor "min".
     """
-
     adj = ssc_dict["adj"]
-    fixef_k = ssc_dict["fixef_k"]
+    fixef_k = ssc_dict["fixef_k"]  # noqa: F841 TODO: is this used?
```
Comment (s3alfisc) [inline pyfixest/utils.py:79]: This is not yet used and still needs to be implemented. This is the main reason why pyfixest standard errors slightly differ from fixest standard errors when there are fixed effects (nested within clusters).

**Review by s3alfisc — APPROVED** [review]

**Comment by s3alfisc** [comment]
The PR looks excellent @NKeleher, thanks so much! I have merged to the dev branch, where I'll run the CI workflows. 

If you would like, I'd be very happy to acknowledge your contributions by adding you to the list of pyfixest authors. Currently it only consists of Styfen (@styfenschaer ) and me, so if you would like to be acknowledged & Styfen is ok with it, I'd add you! =)

**Comment by s3alfisc** [comment]
By the way, with this PR I will also merge the just file - it looks really useful & I should really start using it!

**Pull request merged**